[GH-1762] Bounties screen

alfopisano opened the pull request:
Closes #1762

## Motivation
We needed to create the bounties feature. This affects the pre, in and post game stages. This PR is a first iteration where we are implementing the UI design, animations and the existing backend. Features like the message when you complete the bounty in-game and the status of it on the deathsplash would have its UI implemented but no functionality yet since that would be added on future iterations. 

Since we do not have the bounties loaded on our system yet, we will handle that case and show no bounty in order for the game not to crush.

## Summary of changes
- Added pre-game screen
Reference video: 
https://github.com/lambdaclass/champions_of_mirra/assets/116388676/f6743df5-825d-4f04-9cfb-23bab7a0791a

Result:
https://github.com/lambdaclass/champions_of_mirra/assets/116388676/4a48289e-bab6-4ce1-ab93-8adf8d468b4c

- Added modifications in-game
Reference:
![reference](https://github.com/lambdaclass/champions_of_mirra/assets/116388676/5e065d7f-7b0d-422a-9191-508a0c1e888f)

Result: 
<img width="878" alt="Screenshot 2024-05-24 at 12 14 27" src="https://github.com/lambdaclass/champions_of_mirra/assets/116388676/976578f4-fbdb-4cab-876b-d0445fdb6ed8">
DISCLAIMER: Since we do not have the backend logic to show this message within the match we will just hide this container until we have it. This will be done on a future iteration.

- Modified the deadsplash to show the bounties ready to claim
Reference:
![reference](https://github.com/lambdaclass/champions_of_mirra/assets/116388676/d93b82d4-668d-485b-9733-907839bdd3c3)

Result:
<img width="870" alt="Screenshot 2024-06-05 at 15 40 34" src="https://github.com/lambdaclass/champions_of_mirra/assets/116388676/148719c6-84ca-400b-9d56-44f73546f8ca">
DISCLAIMER: Since we do not have the backend logic to show the status of the bounty and to claim it within the deathsplash we will just hide the this containers until we have it and only show the selected bounty and its reward. This will be done on a future iteration.

## How has this been tested?
To test this you would need to use this backend [PR](https://github.com/lambdaclass/mirra_backend/pull/691). 
- After starting the backend you would need to import the bounties: 
```
GameBackend.CurseOfMirra.Config.import_quest_descriptions_config
``` 
- Also you can play with the amount `bounty_pick_time_ms` on the file `apps/arena/priv/config.json` from the backend in order to make the battle start at the right time. 

After having al this things set up you just need to play and select a bounty. 
Also test without importing the bounties to check that this is handled properly. 

## Checklist
- [ ] I have tested the changes locally.
- [ ] I have tested the whole game after applying the changes, not only the affected areas.
- [ ] I self-reviewed the changes on GitHub, line by line.
- [ ] I have tested the changes in another devices.
  - [ ] Tested in iOS.
  - [ ] Tested in Android.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

tkz00 commented on client/Assets/Scripts/UI/PrepareForBattleAnimations.cs:
```diff
@@ -286,6 +347,16 @@ void GeneratePlayer(int index, Entity player)
 
+    void GenerateBounties()
+    {
+        for (int i = 0; i < 3; i ++) {
```
I feel if the number of bounties that come from the backend vary, the client shouldn't break or be affected by it.

Right now if a list of bounties with less than three elements comes from the backend, then the client breaks, this might never be the case, but I feel it could, in that case this logic should be revised to spawn the same number of bounties in the screen as they come from the backend, let me know what you think.

alfopisano replied:
I feel this is something we need to ask, if the amount of bounties displayed is always the same. But on a general rule I agree with you, it should not break.

tkz00 replied:
Great! We'll leave it like it is, if in the future the number of bounties to pick from becomes variable then we'll iterate on this.

tkz00 reviewed (dismissed):
Looks great! I even tested it on my android device and it worked!

jazimp commented:
small detail: the winner's box isn't the same width as the two containers below it: (reference vs implementation) 
<img width="438" alt="Captura de pantalla 2024-06-05 a la(s) 13 25 18" src="https://github.com/lambdaclass/champions_of_mirra/assets/134973476/de8e6827-fa04-48cf-86cf-16460b4a146f">
<img width="435" alt="Captura de pantalla 2024-06-05 a la(s) 13 25 34" src="https://github.com/lambdaclass/champions_of_mirra/assets/134973476/0bda7b47-ec74-44af-b661-4dffc32a94ab">

jazimp commented:
niiice! visual approval ✅

tkz00 reviewed (approved):
Great work! Also looks good on my android device!

alfopisano commented:
Wee need to handle the case where the backend sends an empty list of bounties, so that it does not crush while we do not have bounties loaded on our system.

ncontinanza reviewed (approved):
LGTM!

tkz00 reviewed (changes requested):
1. `BOUNTIES_PICK_TIME` is defined in uppercase and separated by underscores, like a constant, but it isn't, it could be renamed to `bountiesPickTime`, right? Also I believe it could be move inside the `BountiesCountdown` method, since it's not used anywhere else.

2 . I feel like the field `id` in `Bounty.cs` can be accessed already from the `bountyInfo` field.

3. This is just a question, why do you use an `Event Trigger` component inside the Bounty prefab, instead of just using the already existing `MM Touch Button`?

The rest works amazing, tested on my android.

tkz00 reviewed (approved):
Great! The issues were solved and it works on my android device!

AngieDutra commented on client/Assets/Images/UI/Buttons/button 1.png:
What role does this button have? I think we could have more descriptive names for assets and avoid numbers

AngieDutra reviewed (commented):
Awesome, I just have a couple questions:
- Why were the font assets changed?
![Screenshot 2024-06-18 at 10 51 04](https://github.com/lambdaclass/champions_of_mirra/assets/82987608/d9dcf4a4-2ef6-43d9-8998-658d46604baa)
- How do you feel the counter number? In my opinion the fade out transition could be a bit quicker and have a bit more time to be in display, wdyt?

alfopisano commented:
> Awesome, I just have a couple questions:
> 
> * Why were the font assets changed?
>   ![Screenshot 2024-06-18 at 10 51 04](https://private-user-images.githubusercontent.com/82987608/340703826-d9dcf4a4-2ef6-43d9-8998-658d46604baa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3MjQyMjgsIm5iZiI6MTcxODcyMzkyOCwicGF0aCI6Ii84Mjk4NzYwOC8zNDA3MDM4MjYtZDlkY2Y0YTQtMmVmNi00M2Q5LTg5OTgtNjU4ZDQ2NjA0YmFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjE4VDE1MTg0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWUxYjhjYTE4NWEzM2YxMDMxMzUwYzBkMGVkOTU1YjAzN2Q0N2MzYzljNjcwYjMwOTJlNzE0MTFmOWI5YjdmMGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gtdfpFf99KozqbqwcxYbZmbZpzdvsbrr_RC94CShf1o)
> * How do you feel the counter number? In my opinion the fade out transition could be a bit quicker and have a bit more time to be in display, wdyt?

I added those fonts to add the shadow to the SourceSans3 Semibold font, as done for the bold one (named SourceSans3 black with the shadow). I don’t know if it is the best approach, wdyt?

About the countdown I agree with you! I would do those changes.

AngieDutra commented:
Oh I see. I think if we have semibold text with shadows and without shadows maybe it should be clarified in the name, wdyt?

AngieDutra reviewed (dismissed):
I have the following issues when running the game 
<img width="817" alt="Screenshot 2024-06-24 at 14 40 11" src="https://github.com/lambdaclass/champions_of_mirra/assets/82987608/0d77521c-d73b-49c8-b4f9-6e49711a9d4e">
<img width="885" alt="Screenshot 2024-06-24 at 14 40 18" src="https://github.com/lambdaclass/champions_of_mirra/assets/82987608/e009aeb7-7d5f-4655-adf6-86d9603e091b">

I also see files modified in my client once I run the game

<img width="467" alt="Screenshot 2024-06-24 at 14 41 50" src="https://github.com/lambdaclass/champions_of_mirra/assets/82987608/978f05c9-c54e-4c18-b6d0-65ecb805bfa1">

AngieDutra reviewed (changes requested):
I think the numbers may not be centered. If they are maybe adjust them so the eye is tricked into thinking it so too
<img width="200" alt="Screenshot 2024-06-24 at 16 41 34" src="https://github.com/lambdaclass/champions_of_mirra/assets/82987608/270fb9e0-1a84-4d3a-a27f-771bad101d70">

Saw in the reference vid that the text and counter also have animations that are missing in the implementation

Questions:
- Is it worth having a claim button? Is there any case where we don't want to claim our achievements?
<img width="1678" alt="Screenshot 2024-06-24 at 16 47 23" src="https://github.com/lambdaclass/champions_of_mirra/assets/82987608/8ce1ba18-27ef-4b99-9598-fb8530d9f71c">
- I haven't seen the coin animation in the reference vid. Also saw the coin changed sizes. Is this PR suppose to change that animation and sizes too or the coin animation should stay as it is?

tkz00 commented:
https://github.com/lambdaclass/champions_of_mirra/assets/67983386/027de1b0-40e4-4f28-a2b5-8867851b0011

Also I saw that the new UI cover's Valtimer's face on it's victory animation, it might be for some specific screen resolutions.

alfopisano commented:
> Screen_Recording_20240624_170627_client.mp4 
> Also I saw that the new UI cover's Valtimer's face on it's victory animation, it might be for some specific screen resolutions.

I think this should be something to discuss with @jazimp

jazimp commented:
Visual approval ✅

jazimp commented:
We discussed it with Alfo, and we believe that, for now, the best option is to hide the character rank information. When the time comes, we'll review and adjust it

alfopisano commented:
As discussed with @jazimp the coin animation is ok as I implemented it. Related to the other changes, I centered the counter, added the missing animations for the counter and the title and added the claim button as a "dummy" (this was discussed with Marco)

AngieDutra commented:
Awesome! Could you send an updated version of how it should look so I can review if this what I see is accurate?

alfopisano commented:
Uploading Screen Recording 2024-06-25 at 18.05.12.mov…
<img width="881" alt="Screenshot 2024-06-25 at 18 06 29" src="https://github.com/lambdaclass/champions_of_mirra/assets/116388676/ac78466b-8ac8-44fd-8abc-7f30785f9002">
<img width="877" alt="Screenshot 2024-06-25 at 18 08 34" src="https://github.com/lambdaclass/champions_of_mirra/assets/116388676/271e2ec5-e41d-48b6-bff1-8eb5ab9bb2d9">

tkz00 reviewed (approved):
Works and looks good on my android!

AngieDutra reviewed (approved):
Tested on desktop and Andoird build and is working as expected
